File: backend/src/test/java/com/chominjungum/hangul/DictationCompareTest.java
package com.chominjungum.hangul;

import static org.assertj.core.api.Assertions.assertThat;

import org.junit.jupiter.api.Test;

/** Dart·TS 구현과 같은 기대치를 유지한다. */
class DictationCompareTest {

    @Test
    void 완전_일치() {
        var r = DictationCompare.score("가나", "가나");
        assertThat(r.correctCount()).isEqualTo(2);
        assertThat(r.scorePercent()).isEqualTo(100);
    }

    @Test
    void 부분_일치() {
        var r = DictationCompare.score("안녕하세요", "안녕하세오");
        assertThat(r.correctCount()).isEqualTo(4);
        assertThat(r.totalCount()).isEqualTo(5);
        assertThat(r.scorePercent()).isEqualTo(80);
        assertThat(r.matches().get(4).mismatchReason()).contains("글자 불일치");
    }

    @Test
    void 받침_차이를_잡는다() {
        var r = DictationCompare.score("간", "가");
        assertThat(r.correctCount()).isZero();
    }

    @Test
    void 답이_짧으면_누락() {
        var r = DictationCompare.score("안녕하세요", "안녕");
        assertThat(r.correctCount()).isEqualTo(2);
        assertThat(r.matches().get(2).mismatchReason()).isEqualTo("글자 누락");
    }

    @Test
    void 초과_입력은_별도_match() {
        var r = DictationCompare.score("가", "가나다");
        assertThat(r.totalCount()).isEqualTo(1);
        assertThat(r.hasExtraInput()).isTrue();
        assertThat(r.matches()).hasSize(3);
        assertThat(r.matches().get(1).mismatchReason()).isEqualTo("불필요한 글자");
    }

    @Test
    void 특수문자와_숫자도_채점한다() {
        var r = DictationCompare.score("1반!", "1반?");
        assertThat(r.correctCount()).isEqualTo(2);
        assertThat(r.matches().get(2).mismatchReason()).contains("기호·숫자 불일치");
    }

    @Test
    void 허용_밖_문자는_오류() {
        assertThat(DictationCompare.score("abc", "가").error()).contains("정답이");
        assertThat(DictationCompare.score("가", "abc").error()).contains("답안이");
    }

    @Test
    void 요약은_인덱스와_정오만_남긴다() {
        var summaries = DictationCompare.score("가나", "가다").matchSummaries();
        assertThat(summaries).hasSize(2);
        assertThat(summaries.get(0)).containsEntry("i", 0).containsEntry("ok", true);
        assertThat(summaries.get(1)).containsEntry("ok", false).containsKey("why");
    }
}

File: packages/hangul-core-ts/test/dictation-compare.test.ts
import { describe, expect, it } from 'vitest';

import { score, scorePercent, summarizeMatches } from '../src/dictation-compare.js';

/** Dart `DictationCompare` 테스트와 같은 기대치를 유지한다. */
describe('DictationCompare (Dart 구현과 동일 규칙)', () => {
  it('완전 일치', () => {
    const r = score('가나', '가나');
    expect(r.correctCount).toBe(2);
    expect(r.totalCount).toBe(2);
    expect(r.ratio).toBe(1);
    expect(scorePercent(r)).toBe(100);
  });

  it('부분 일치 — 글자별 정오와 사유', () => {
    const r = score('가나', '가다');
    expect(r.correctCount).toBe(1);
    expect(r.totalCount).toBe(2);
    expect(r.matches[1]!.isCorrect).toBe(false);
    expect(r.matches[1]!.mismatchReason).toContain('글자 불일치');
  });

  it('받침 차이를 잡아낸다', () => {
    const r = score('간', '가');
    expect(r.correctCount).toBe(0);
    expect(r.matches[0]!.mismatchReason).toContain('글자 불일치');
  });

  it('답이 짧으면 누락으로 센다', () => {
    const r = score('안녕하세요', '안녕');
    expect(r.correctCount).toBe(2);
    expect(r.totalCount).toBe(5);
    expect(r.matches[2]!.mismatchReason).toBe('글자 누락');
    expect(scorePercent(r)).toBe(40);
  });

  it('초과 입력은 별도 match 로 덧붙고 플래그가 선다', () => {
    const r = score('가', '가나다');
    expect(r.totalCount).toBe(1);
    expect(r.correctCount).toBe(1);
    expect(r.hasExtraInput).toBe(true);
    expect(r.matches).toHaveLength(3);
    expect(r.matches[1]!.mismatchReason).toBe('불필요한 글자');
  });

  it('특수문자·숫자도 채점 대상', () => {
    const r = score('1반!', '1반?');
    expect(r.correctCount).toBe(2);
    expect(r.matches[2]!.mismatchReason).toContain('기호·숫자 불일치');
  });

  it('허용 밖 문자는 오류로 반환한다', () => {
    expect(score('abc', '가').error).toContain('정답이');
    expect(score('가', 'abc').error).toContain('답안이');
  });

  it('업싱크용 요약은 인덱스·정오·사유만 남긴다', () => {
    const summary = summarizeMatches(score('안녕하세요', '안녕하세오'));
    expect(summary).toHaveLength(5);
    expect(summary[0]).toEqual({ i: 0, ok: true });
    expect(summary[4]!.ok).toBe(false);
    expect(summary[4]!.why).toContain('글자 불일치');
  });
});
